### PR TITLE
Fix Migration 145 on MSSQL if varchar is changed to nvarchar

### LINKED ISSUE #12441: Issue with :latest docker image with MSSQL database.
- Gitea version (or commit ref):
- Git version:
Docker Latest (d523f424804cf80b3be0c0244241da7cc5ac82023c5daa04bd46a0cb38718e12)
- Operating system:
Linux/Docker
- Database (use `[x]`):
  - [ ] PostgreSQL
  - [ ] MySQL
  - [ X] MSSQL
  - [ ] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [ ] Yes (provide example URL)
  - [ ] No
  - [ X ] Not relevant
- Log gist:
https://gist.github.com/onemustpersist/6e8a0c5735c8c9386bd391c97b8968b9
## Description

Fails to start after upgrade. There is a database roll out issue. Using specific 1.12.3 allows git to start, but UI fails due to incorrect SQL query:
> 
2020/08/06 06:59:39 routers/user/home.go:60:getDashboardContextUser() [E] GetOrganizations: mssql: Column 'user.keep_activity_private' is invalid in the select list because it is not contained in either an aggregate function or the GROUP BY clause.

## Screenshots
Not applicable

### LINKED ISSUE #13013: Gitea crash when indexer is run (database error, data too long for column language)
- Gitea version (or commit ref): Gitea version 1.12.5 built with GNU Make 4.1, go1.14.9 : bindata, sqlite, sqlite_unlock_notify
- Git version: 2.28.0
- Operating system: Ubuntu 18.04.5 LTS
- Database (use `[x]`):
  - [ ] PostgreSQL
  - [x] MySQL
  - [ ] MSSQL
  - [ ] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [ ] Yes (provide example URL)
  - [x] No
  - [ ] Not relevant
- Log gist:
```
./gitea.log:2562:2020/10/02 09:12:07 ...dexer/stats/queue.go:24:handle() [E] stats queue indexer.Index(24) failed: Error 1406: Data too long for column 'language' at row 1
```

## Description
I've upgraded Gitea today the usual way by stopping the service, replacing binary and starting the service again. (From 1.12.4 to 1.12.5)

Gitea crashes when indexer runs. I have no idea why this happened. Maybe failed database migration on my end? That's the only relevant log I could find above.

I've disabled repo indexer and things started working again.

### PR DESCRIPTION
**Pull request by zeripath**
Some versions of MSSQL will fail on migration v145 because there is a constraint failure.

This PR drops the constraints on the language column prior to its alteration and then recreates them with sync.

Fix #12441
Fix #13013

Signed-off-by: Andrew Thornton <art27@cantab.net>

### CODE REVIEW

File: models/migrations/v145.go [hunk 1]
```diff
@@ -47,9 +47,29 @@ func increaseLanguageField(x *xorm.Engine) error {
...
+		for _, constraint := range constraints {
+			if _, err := sess.Exec(fmt.Sprintf("ALTER TABLE `language_stat` DROP CONSTRAINT `%s`", constraint)); err != nil {
+				return fmt.Errorf("Drop table `language_stat` constraint `%s`: %v", constraint, err)
+			}
+		}
 		if _, err := sess.Exec(fmt.Sprintf("ALTER TABLE language_stat ALTER COLUMN language %s", sqlType)); err != nil {
 			return err
 		}
+		// Finally restore the constraint
+		if err := x.Sync2(new(LanguageStat)); err != nil {
```
Comment (zeripath) [inline models/migrations/v145.go]:
```suggestion
		if err := sess.Sync2(new(LanguageStat)); err != nil {
```

File: models/migrations/v145.go [hunk 2]
```diff
@@ -47,9 +47,29 @@ func increaseLanguageField(x *xorm.Engine) error {
 			return err
 		}
 	case setting.Database.UseMSSQL:
+		// Yet again MSSQL just has to be awkward.
+		// Here we have to drop the constraints first and then rebuild them
```
Comment (6543) [inline models/migrations/v145.go:51]: I think this is more a thing vor xorm?

Comment (zeripath) [reply models/migrations/v145.go:51]: But Xorm doesn't handle it. Therefore we have to. 

Currently people on certain earlier versions of mssql cannot run Gitea docker. We need to fix this.

**Review by 6543 — APPROVED** [review]

**Review by lafriks — APPROVED** [review]

**Comment by zeripath** [comment]
Make lg-tm work

File: models/migrations/v145.go [hunk 3]
```diff
@@ -47,9 +47,29 @@ func increaseLanguageField(x *xorm.Engine) error {
...
+		for _, constraint := range constraints {
+			if _, err := sess.Exec(fmt.Sprintf("ALTER TABLE `language_stat` DROP CONSTRAINT `%s`", constraint)); err != nil {
+				return fmt.Errorf("Drop table `language_stat` constraint `%s`: %v", constraint, err)
+			}
+		}
 		if _, err := sess.Exec(fmt.Sprintf("ALTER TABLE language_stat ALTER COLUMN language %s", sqlType)); err != nil {
 			return err
 		}
+		// Finally restore the constraint
+		if err := sess.Sync2(new(LanguageStat)); err != nil {
```
Comment (zeripath) [inline models/migrations/v145.go]:
```suggestion
		if err := sess.CreateIndexes(new(LanguageStat)); err != nil {
```

File: models/migrations/v145.go [hunk 4]
```diff
@@ -47,9 +47,29 @@ func increaseLanguageField(x *xorm.Engine) error {
...
+		for _, constraint := range constraints {
+			if _, err := sess.Exec(fmt.Sprintf("ALTER TABLE `language_stat` DROP CONSTRAINT `%s`", constraint)); err != nil {
+				return fmt.Errorf("Drop table `language_stat` constraint `%s`: %v", constraint, err)
+			}
+		}
 		if _, err := sess.Exec(fmt.Sprintf("ALTER TABLE language_stat ALTER COLUMN language %s", sqlType)); err != nil {
 			return err
 		}
+		// Finally restore the constraint
+		if err := sess.CreateIndexes(new(LanguageStat)); err != nil {
```
Comment (zeripath) [inline models/migrations/v145.go]:
```suggestion
		if err := sess.CreateUniques(new(LanguageStat)); err != nil {
```

**Comment by lafriks** [comment]
Still fails with error:
```
mssql: The operation failed because an index or statistics with name 'UQE_language_stat_s' already exists on table 'language_stat'.
```

**Comment by zeripath** [comment]
It looks like: 

```sql
"SELECT Name FROM SYS.DEFAULT_CONSTRAINTS WHERE " +
				"PARENT_OBJECT_ID = OBJECT_ID('language_stat') AND " +
				"PARENT_COLUMN_ID IN (SELECT column_id FROM sys.columns " +
				"WHERE lower(NAME) = 'language' AND " +
				"object_id = OBJECT_ID('language_stat'))").Find(&constraints)
```

isn't dropping the unique constraint. Which is weird

**Comment by zeripath** [comment]
OK I'm going to try a different query...

```sql
SELECT i.name AS Name
	FROM sys.indexes i INNER JOIN sys.index_columns ic
		ON i.index_id = ic.index_id AND i.object_id = ic.object_id
	INNER JOIN sys.tables AS t 
		ON t.object_id = i.object_id
	INNER JOIN sys.columns c
		ON t.object_id = c.object_id AND ic.column_id = c.column_id
	WHERE t.name = 'language_stat' AND c.name = 'language'
```

**Comment by zeripath** [comment]
OK that finally appears to work.

**Pull request merged**